src/proxy.rs

```diff
@@ -526,6 +530,60 @@ pub fn guess_inbound_service(
         .map(ServiceDescription::from)
 }
 
+// Checks if the connection's source identity is the identity for the upstream's waypoint
+async fn check_from_waypoint(
+    state: DemandProxyState,
+    upstream: &Workload,
+    src_identity: Option<&Identity>,
+) -> bool {
+    check_gateway_address(state, src_identity, upstream.waypoint.as_ref()).await
+}
+
+// Checks if the connection's source identity is the identity for the upstream's network
+// gateway
+async fn check_from_network_gateway(
+    state: DemandProxyState,
+    upstream: &Workload,
+    src_identity: Option<&Identity>,
+) -> bool {
+    check_gateway_address(state, src_identity, upstream.network_gateway.as_ref()).await
+}
+
+// Check if the source's identity matches any workloads that make up the given gateway
+// TODO: This can be made more accurate by also checking addresses.
+async fn check_gateway_address(
+    state: DemandProxyState,
+    src_identity: Option<&Identity>,
+    gateway_address: Option<&GatewayAddress>,
+) -> bool {
+    let Some(src_identity) = src_identity else {
+        return false;
+    };
+    if let Some(gateway_address) = gateway_address {
+        let from_gateway = match state.fetch_destination(&gateway_address.destination).await {
+            Some(Address::Workload(wl)) => &wl.identity() == src_identity,
+            Some(Address::Service(svc)) => {
+                for (_ep_uid, ep) in svc.endpoints.iter() {
+                    // fetch workloads by workload UID since we may not have an IP for an endpoint (e.g., endpoint is just a hostname)
+                    if state
+                        .fetch_workload_by_uid(&ep.workload_uid)
+                        .await
+                        .map(|w| w.identity())
+                        .as_ref()
+                        == Some(src_identity)
+                    {
+                        return true;
```

Comment:
You're reading it right, but also this code isn't new. I moved it here because it's now also used in outbound. 

This is usually used for checking the source of traffic, we could expand it to include an address.

Reply:
I ended up changing this to check identity and IP. 

> require at-most-one backing workload for gateway services

Sort of.. the from_waypoint check in outbound is deciding if we need to traverse the waypoint for sandwich. This is only relevant for sandwiched gateways which have workloads.

The from_waypoint check in inbound will is used to prevent bypass. For now this change does break waypoints that don't have a Workload, but that will eventually be replaced by RBAC rules. 

We could use different logic in each place though.

Reply:
The reason for the change is that some tests were breaking. I assume the tests had things sharing identities.